Add API endpoints and Jetpack Backup package for managing Helper Scripts

thingalon opened the pull request:
Project details: pa0RFL-oP-p2

Prototype P2 post: pa0RFL-ra-p2
Related wpcom patch: D34445-code
Related transport patch: 975-gh-jetpack-backups

This PR adds support for Jetpack Backup Helper Scripts. It includes two new API endpoints; one for installing a Jetpack Backup Helper Script, and one for deleting them. It also adds a scheduled action to cleanup any expired Helper Scripts after any get installed. 

#### Changes proposed in this Pull Request:
- New module; Jetpack Backup - for helper methods for the backup system. For now, it only contains one class; a class for managing Helper Scripts. 
- New API endpoints; one for installing a Jetpack Backup Helper Script, and one for deleting one.

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
This adds support for Zero-Config Backups to Jetpack. 

#### Testing instructions:
See instructions on the transport patch: 975-gh-jetpack-backups

#### Proposed changelog entry for your changes:
* Support for Jetpack Backups with simpler configuration.

Commit:
Add API endpoints and Jetpack Backup package for managing Helper Scripts

seear committed:
Add new endpoint files to phpcs whitelist

New files need linting.

seear commented:
Rebased to master.

enejb commented on packages/backup/composer.json:
```diff
@@ -0,0 +1,17 @@
+	"require": {},
+	"autoload": {
+		"files": [
+			"actions.php"
```
Would this file be always autoloaded?
Would't we have conflict here? where we are defining the same function multiple times 
if this package were to be included in multiple plugins?

thingalon replied:
You are right. I have removed the one and only function from actions.php, as per your other suggestion. :)

enejb commented on packages/backup/actions.php:
```diff
@@ -0,0 +1,18 @@
+function jetpack_backup_cleanup_helper_scripts() {
+	Automattic\Jetpack\Backup\Helper_Script_Manager::cleanup_expired_helper_scripts();
+}
+add_action( 'jetpack_backup_cleanup_helper_scripts', 'jetpack_backup_cleanup_helper_scripts' );
```
Can we not call the method directly? Instead of wrapping in into a function?

thingalon replied:
Yes, you are right. This is redundant. I have removed the function, and made the action directly call into Helper_Script_Manager.

... I didn't realize autoload worked with hooks, but there you go. Tested, and it works. Thanks for the suggestion! :)

Commit:
[not verified] Call Automattic\Jetpack\Backup\Helper_Script_Manager::cleanup_expired_helper_scripts directly from the action hook, instead of via a function

jeherve commented:
Related discussion: p9ue0y-81-p2

jeherve commented on packages/backup/src/Helper_Script_Manager.php:
```diff
@@ -0,0 +1,238 @@
+ */
+class Helper_Script_Manager {
+
+	const RELATIVE_INSTALL_LOCATIONS = array( '', 'wp-content', 'wp-content/uploads' );
```
Should we make this dynamic to catch folks using custom wp-content locations with `WP_CONTENT_DIR`?

thingalon replied:
Thank you; that is a great suggestion, and a huge improvement. It's also caused me to update the Helper Script template to make it accept a configurable WordPress install location, instead of assuming it's installed in a subdirectory of it. :)

jeherve commented on packages/backup/src/Helper_Script_Manager.php:
```diff
@@ -0,0 +1,238 @@
+
+		// Generate a random filename, avoid clashes.
+		$max_attempts = 5;
+		for ( $attempt = 0; $attempt < $max_attempts; $attempt++ ) {
```
The multiple attempts are to cover potential filename clashes? When would that happen since our file is prefixed with `jp-helper`?

thingalon replied:
It's possible for multiple Transport Servers to attempt to communicate with the same site, and as a result multiple jp-helpers can get installed. Clashes are unlikely, but this ensures they never become a problem.

jeherve reviewed (changes requested):
I can't really test this without a VP sandbox, so my comments will mostly be nitpicking.

********

What do you think about doing some cleanup in `uninstall.php` to make sure we always clean after ourselves if someone deletes the plugin from their site?

********

There are quite a few PHPCS errors and warning that will stop you from committing changes to those files. Could you fix those errors?

Commit:
[not verified] On uninstallation, delete all Helper Scripts

Commit:
Switch to using WP_Filesystem instead of file_get_contents and file_put_contents

Commit:
[not verified] Rename new api endpoint filenames in compliance with phpcs requirements

Commit:
[not verified] Rename new api endpoint filenames in compliance with phpcs requirements

Commit:
Document API endpoints as per linter requirements

Commit:
Use use instead of bulky call syntax

thingalon commented on json-endpoints/jetpack/class.jetpack-json-api-delete-backup-helper-script.php:
```diff
@@ -0,0 +1,33 @@
+	}
+
+	protected function delete() {
+		$this->result = Automattic\Jetpack\Backup\Helper_Script_Manager::delete_helper_script( $this->script_path );
```
Change applied. :) 

I avoided using use, to make sure that the Helper_Script_Manager wasn't loaded until the very last minute, but it does make the code very ugly.

thingalon commented on json-endpoints/jetpack/class.jetpack-json-api-delete-backup-helper-script.php:
```diff
@@ -0,0 +1,33 @@
+
+	protected function delete() {
+		$this->result = Automattic\Jetpack\Backup\Helper_Script_Manager::delete_helper_script( $this->script_path );
+		Automattic\Jetpack\Backup\Helper_Script_Manager::cleanup_expired_helper_scripts();
```
👍

thingalon commented on json-endpoints/jetpack/class.jetpack-json-api-delete-backup-helper-script.php:
```diff
@@ -0,0 +1,33 @@
+<?php
+
```
👍

thingalon commented on json-endpoints/jetpack/class.jetpack-json-api-install-backup-helper-script.php:
```diff
@@ -0,0 +1,40 @@
+	}
+
+	protected function install() {
+		$this->result = Automattic\Jetpack\Backup\Helper_Script_Manager::install_helper_script( $this->helper_script );
```
👍

thingalon commented on json-endpoints/jetpack/class.jetpack-json-api-install-backup-helper-script.php:
```diff
@@ -0,0 +1,40 @@
+
+	protected function install() {
+		$this->result = Automattic\Jetpack\Backup\Helper_Script_Manager::install_helper_script( $this->helper_script );
+		Automattic\Jetpack\Backup\Helper_Script_Manager::cleanup_expired_helper_scripts();
```
👍

thingalon commented on json-endpoints/jetpack/class.jetpack-json-api-install-backup-helper-script.php:
```diff
@@ -0,0 +1,40 @@
+<?php
+
```
👍

Commit:
Change Jetpack Backups -> Jetpack Backup, also clean up formatting

Commit:
Tidy up alignment of => markers

thingalon commented on json-endpoints/jetpack/json-api-jetpack-endpoints.php:
```diff
@@ -1262,3 +1262,55 @@
+// POST /sites/%s/install-backup-helper-script
+require_once( $json_jetpack_endpoints_dir . 'class.jetpack-json-api-install-backup-helper-script.php' );
+new Jetpack_JSON_API_Install_Backup_Helper_Script_Endpoint( array(
+	'description'    => 'Setup a Helper Script, to allow Jetpack Backups to connect to this site',
```
Thanks for picking up on this. :) Fixed.

Commit:
Use a more explicit header comment; making it clearer that this is related to Jetpack Backup

thingalon commented on packages/backup/src/Helper_Script_Manager.php:
```diff
@@ -0,0 +1,238 @@
+
+	const RELATIVE_INSTALL_LOCATIONS = array( '', 'wp-content', 'wp-content/uploads' );
+	const TEMP_DIRECTORY             = 'jetpack-temp';
+	const HELPER_HEADER              = "<?php /* JPR Helper Script */\n";
```
Good point. I've updated it to "Jetpack Backup", as JPR is not identifiable externally. We can add more comments after this to the Helper Script, describing what it is in more detail - but that is up to the Transport Server providing the Helper.

thingalon commented on json-endpoints/jetpack/class.jetpack-json-api-install-backup-helper-script.php:
```diff
@@ -0,0 +1,40 @@
+class Jetpack_JSON_API_Install_Backup_Helper_Script_Endpoint extends Jetpack_JSON_API_Endpoint {
+	// /sites/%s/install-backup-helper-script
+
+	protected $needed_capabilities = array(); // This endpoint is only accessible from Jetpack Backups
```
👍  Taken care of in previous tidy-ups.

tyxla commented on json-endpoints/jetpack/class-jetpack-json-api-install-backup-helper-script-endpoint.php:
```diff
@@ -0,0 +1,85 @@
+
+use Automattic\Jetpack\Backup\Helper_Script_Manager;
+
+class Jetpack_JSON_API_Install_Backup_Helper_Script_Endpoint extends Jetpack_JSON_API_Endpoint { 
```
Have you considered creating those endpoints as v2 endpoints (like [this one](https://github.com/Automattic/jetpack/blob/3ed3de86a58730377bbc29f8b7df7aa7729f28ce/_inc/lib/core-api/wpcom-endpoints/memberships.php))? I thought we try to create new endpoints with this mechanism, rather than the legacy v1 one.

Just in case it's helpful, this post is a good place to read more about the different types of endpoints: PCYsg-aqU-p2

tyxla commented on packages/backup/composer.json:
```diff
@@ -0,0 +1,17 @@
+{
```
It's really cool we're considering to add this functionality as a new package. I, however, was wondering if there was any special purpose behind that - are we planning to use this package anywhere else?

thingalon replied:
I'm not sure if this will be useful in other projects down the line, but keeping it in a package: 
1) Keeps it contained in a logical manner, and
2) Would make it easier to split this out into a separate plugin, if we decide to down the line (as this is something we've discussed as a possibility).

tyxla replied:
Sounds like a great reason to keep it in a package already, thanks for elaborating! 😉

tyxla commented on packages/backup/src/Helper_Script_Manager.php:
```diff
@@ -0,0 +1,322 @@
+		for ( $attempt = 0; $attempt < $max_attempts; $attempt++ ) {
+			$file_key           = wp_generate_password( 10, false );
+			$relative_file_path = trailingslashit( $relative_temp_dir ) . 'jp-helper-' . $file_key . '.php';
+			$absolute_file_path = trailingslashit( ABSPATH ) . $relative_file_path;
```
Should this be `\ABSPATH` because we're in a different namespace from the global one here?

thingalon replied:
Yes, probably better to be explicit. This instance of ABSPATH was refactored out based on previous feedback just as your comments were added -- but I've added \\ to the new instance of ABSPATH (and similar constants) in the latest version. 

Thanks for the feedback! :)

tyxla replied:
Thank you! Also, noting that we should be able to use the `constants` package to safely check and retrieve constants.

tyxla commented on packages/backup/src/Helper_Script_Manager.php:
```diff
@@ -0,0 +1,322 @@
+	 * @param string $path Path to Helper Script to delete.
+	 * @return boolean     True if the file is deleted (or does not exist).
+	 */
+	public static function delete_helper_script( $path ) {
```
I wonder how many of these methods really have to be static.

thingalon replied:
They're all static, because there is no instance of a `Helper_Script_Manager` object. I followed a similar pattern to many of the classes in the Jetpack Sync package, which are logical lumps of static methods which are easy to autoload. 

Would it be preferable to have each of these as a separate function just in the `Automattic\Jetpack\Backup` namespace?

tyxla replied:
I see. If you feel these should be static and you don't consider having an instance of this object, I think they're good already as they are 👍

tyxla commented on packages/backup/src/Helper_Script_Manager.php:
```diff
@@ -0,0 +1,322 @@
+	private static function put_contents( $file_path, $contents ) {
+		global $wp_filesystem;
+
+		if ( ! function_exists( 'WP_Filesystem' ) ) {
```
Should we prefix this one with `\\` as it's in the global namespace?

thingalon replied:
Good point. Added. Thanks :)

tyxla reviewed (commented):
I love how readable and both self-documented and well-documented this code is. Thanks for the great work!

I haven't tested it, but I've left some drive-by comments and questions in case they can be useful.

Commit:
Use WP_CONTENT_DIR, WP_CONTENT_URL and wp_upload_dir to determine wp-content and upload locations without making assumptions

thingalon commented:
Thanks for the review @jeherve 

> What do you think about doing some cleanup in uninstall.php to make sure we always clean after ourselves if someone deletes the plugin from their site?

Good idea; added. Thanks! :)

> There are quite a few PHPCS errors and warning that will stop you from committing changes to those files. Could you fix those errors?

Cleaned up! Thanks :)

Commit:
Explicitly use constants and WordPress methods from the global namespace, remove two unused variables

Commit:
Escape the backslash in function_exists for WP_Filesystem to actually check for it in the global namespace.

jeherve commented:
Had to rebase to fix conflicts with `master`.

seear commented:
Testing this, I find I still need to use the old `<?php /* JPR Helper Script */` header. Is there some extra step necessary to use the latest `automattic/jetpack-backup` package on the Jetpack site?

seear commented:
Rebased to `master`.

seear commented:
> Is there some extra step necessary to use the latest automattic/jetpack-backup package on the Jetpack site?

Failing Travis build because of the wrong filenames in whitelist was the problem.

seear committed:
Correct endpoint filenames in phpcs whitelist

seear committed:
Missing comma

seear commented:
I've re-tested using latest patches D34445-code and 975-gh-jetpack-backups. The helper script uploads and a backup runs.

jeherve reviewed (approved):
I can't test all of this since I don't have a VP sandbox, but the Jetpack side looks good. I'm merging this now to make testing easier.

Conversation at p9ue0y-81-p2 is still ongoing so keeping an eye on that too.

D34452-code, D34445-code, and 975-gh-jetpack-backups will also need to be merged.
